spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -663,6 +663,9 @@ public void run() {
 				KafkaMessageListenerContainer.this.stop();
 			}
 			this.monitorTask.cancel(true);
+			if (this.taskScheduler instanceof ThreadPoolTaskScheduler) {
+				((ThreadPoolTaskScheduler) this.taskScheduler).destroy();
```

Comment:
I see why you do this:
```
if (this.containerProperties.getScheduler() != null) {
	this.taskScheduler = this.containerProperties.getScheduler();
}
else {
	ThreadPoolTaskScheduler threadPoolTaskScheduler = new ThreadPoolTaskScheduler();
	threadPoolTaskScheduler.initialize();
	this.taskScheduler = threadPoolTaskScheduler;
}
```
So, we create an internal non-managed instance if it isn't provided via `ContrainerProperties`.

Well, that's indeed a good catch.
And thank you very much for the contribution on the matter!

Only the problem that we can't do that unconditionally.

In other places we have some `boolean` flag to control injection/internal instance. Something like `taskSchedulerExplicitlySet`. And call this `destroy()` only if it is `false`.